File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveFunctionReturnTypeNullableUnionName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node) => {
    if (
        node.parent
        && ts.isUnionTypeNode(node.parent)
        && node.parent.types.length > 1
        && node.parent.types[1].kind === ts.SyntaxKind.UndefinedKeyword

        && node.parent.parent
        && ts.isFunctionDeclaration(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.name)
        && (node.parent.parent.type === node.parent)
    ) {
        const parentName = node.parent.parent.name.text

        return `${karakum.capitalize(parentName)}Result`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/wrappersgenerator/node/nameResolvers/resolveFunctionReturnTypeNullableUnionName.kt
package wrappersgenerator.node.nameResolvers

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.*

fun resolveFunctionReturnTypeNullableUnionName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService(typeScriptServiceKey))

    val union = ensureNotNull(typeScriptService.getParent(node))
    ensure(isUnionTypeNode(union))
    ensure(union.types.asArray().size > 1)

    val unionSecondType = ensureNotNull(union.types.asArray().getOrNull(1))
    ensure(unionSecondType.kind == SyntaxKind.UndefinedKeyword)

    val function = ensureNotNull(typeScriptService.getParent(union))
    ensure(isFunctionDeclaration(function))
    ensure(function.type === union)
    val parentName = ensureNotNull(function.name).text

    "${parentName.replaceFirstChar { it.titlecase() }}Result"
}
